


#"Reference and variant allele read counts were extraced from the bams files for germline polymorphic 
#sites cataloagues in the dvSNP and 1000genome databases."
## snp-pileup
#This application will, given a VCF file containing SNP locations, output for
#each SNP the counts of the reference nucleotide, alternative nucleotide,
#errors, and deletions. These counts can then be used in facets.


#Use dbsnp151? #add gnomAD?

#Used 00-common_all.vcf.gz which was retrieved from NCBI
#I think its all common snps for GRCh37.

#Can I make a uniform input for seqz and facets?
#turn one into the other or vice versa?


snp-pileup -g -q15 -Q20 -P100 -r25,0 vcffile outputfile normalbam tumorbam
