


rule write_pindel_config:
    input:
        tumor_bam=
        normal_bam=
    output:
        "data/work/{targets}/{tumor}/pindel/config.txt"
    shell:
        """
        echo {input.tumor_bam} 250 {tumor} > {output}
        echo {input.normal_bam} 250 {normal} >> {output}
        """

rule run_pindel_byChr:
#by Chr
    input:
        rule.write_pindel_config.output
    output:
        data/work/{targets}/{tumor}/pindel/pindel.chr{chr}_D,
        data/work/{targets}/{tumor}/pindel/pindel.chr{chr}_INV,
        data/work/{targets}/{tumor}/pindel/pindel.chr{chr}_SI,
        data/work/{targets}/{tumor}/pindel/pindel.chr{chr}_TD,
    params:
        ref=config['reference']['fasta'],
        outname="data/work/{targets}/{tumor}/pindel/pindel.chr{chr}"
    threads:
        
    shell:
        "/home/bwubb/software/pindel/pindel -f {params.ref} -T {threads} -i {input} -c {chr} -o {params.outname}"

rule run_pindel2vcf_byChr:
    input:
        rules.run_pindel_byChr.output
    output:
        "data/work/{targets}/{tumor}/pindel/pindel.chr{chr}.vcf"
    params:
        ref=config['reference']['fasta'],
        refkey=config['reference']['key'],
        outname="data/work/{targets}/{tumor}/pindel/pindel.chr{chr}"
    shell:
        "/home/bwubb/software/pindel/pindel2vcf -P {params.outname} -G -r {params.ref} -R {params.refkey} -d 20090201 -c {chr}"
#by chromosome
#~/software/pindel/pindel2vcf -P test-name -G -r /home/bwubb/resources/Genomes/Human/GRCh37/human_g1k_v37.fasta -R GRCh37 -d 20090201 -c 17
#filter tweeking

rule filter_pindel_vcf:
#A lot of potential filtering work.
#Depths, AD,
#split to somatic and germline

rule prepare_pindel_avinput
#seems likely will have write your own avinput
#13 20797176 21105944 0 - comments:
#(chr, start, end, ref, alt, plus optional fields)
#The last variant is a large-scale deletion, but the reference allele is represented by “0”, 
#  eliminating the need to include reference nucleotides on this line.
##Need to look at what input (REF/ALT) works and what does not.

rule annovate_pindel
#table_annovar.pl --buildver hg19 --outfile test-name.annovar --otherinfo --remove --protocol refGene --operation g  test-name.custom.avinput ~/resources/annovar/humandb/
# makes test-name.annovar.hg19_multianno.txt
#can convert to vcf again or make into annovartools like output, converting 0 to "0" or .